src/plugins/shared_dictionaries_plugin/shared_dictionaries_column_processor.cpp

```diff
@@ -0,0 +1,291 @@
+    const auto chunk = table->get_chunk(chunk_id);
+    const auto segment = chunk->get_segment(column_id);
+
+    const auto dictionary_segment = std::dynamic_pointer_cast<DictionarySegment<T>>(segment);
+    if (dictionary_segment && dictionary_segment->uses_dictionary_sharing()) {
+      stats->num_existing_merged_dictionaries++;
+      const auto shared_dictionary = dictionary_segment->dictionary();
+      const auto [iter, inserted] =
+          shared_dictionaries_map.try_emplace(shared_dictionary, std::make_shared<MergePlan>(shared_dictionary));
+      const auto segment_chunk_pair = std::make_pair(dictionary_segment, chunk);
+      const auto merge_plan = iter->second;
+      _add_segment_chunk_pair(*merge_plan, segment_chunk_pair, true);
+    }
+  }
+
+  stats->num_existing_shared_dictionaries = static_cast<uint32_t>(shared_dictionaries_map.size());
+
+  for (const auto& [key, merge_plan] : shared_dictionaries_map) {
+    merge_plans.emplace_back(merge_plan);
+  }
+}
+
+template <typename T>
+std::pair<int32_t, std::shared_ptr<const pmr_vector<T>>>
+SharedDictionariesColumnProcessor<T>::_compare_with_existing_merge_plans(
+    const std::shared_ptr<const pmr_vector<T>> current_dictionary,
+    const std::vector<std::shared_ptr<MergePlan>>& merge_plans, const PolymorphicAllocator<T>& allocator) {
+  auto best_merge_plan_index = -1;
+  auto best_jaccard_index = -1.0;
+  std::shared_ptr<const pmr_vector<T>> best_shared_dictionary = nullptr;
+  const auto merge_plans_size = merge_plans.size();
+  for (auto merge_plan_index = 0ul; merge_plan_index < merge_plans_size; ++merge_plan_index) {
+    const auto merge_plan = merge_plans[merge_plan_index];
+    const auto shared_dictionary = merge_plan->shared_dictionary;
+    auto union_result = std::make_shared<pmr_vector<T>>(allocator);
+    union_result->reserve(std::max(current_dictionary->size(), shared_dictionary->size()));
+    std::set_union(current_dictionary->cbegin(), current_dictionary->cend(), shared_dictionary->cbegin(),
+                   shared_dictionary->cend(), std::back_inserter(*union_result));
+    const auto total_size = current_dictionary->size() + shared_dictionary->size();
+    const auto union_size = union_result->size();
+    const auto jaccard_index = _calc_jaccard_index(union_size, total_size - union_size);
+    if (jaccard_index > best_jaccard_index) {
+      if (_should_merge(jaccard_index, current_dictionary->size(), union_size,
+                        merge_plan->segment_chunk_pairs_to_merge)) {
+        best_merge_plan_index = static_cast<int32_t>(merge_plan_index);
+        best_jaccard_index = jaccard_index;
+        union_result->shrink_to_fit();
+        best_shared_dictionary = union_result;
+      }
+    }
+  }
+
+  return std::make_pair(best_merge_plan_index, best_shared_dictionary);
+}
+
+template <typename T>
+std::shared_ptr<const pmr_vector<T>> SharedDictionariesColumnProcessor<T>::_compare_with_previous_dictionary(
```

Comment:
That does not only compare with the previous dictionary but also actually do the merge, doesn't it?

Reply:
We changed the name to `_union_with... `

Fixed in https://github.com/hyrise/hyrise/pull/2394/commits/7f88317d0867f9bfd92688f5061daea73a39be35